example/src/main/java/io/netty/example/stomp/websocket/StompChatHandler.java

```diff
@@ -0,0 +1,193 @@
+/*
+ * Copyright 2020 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.example.stomp.websocket;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.ChannelFutureListener;
+import io.netty.channel.ChannelHandler.Sharable;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.SimpleChannelInboundHandler;
+import io.netty.handler.codec.DecoderResult;
+import io.netty.handler.codec.stomp.DefaultStompFrame;
+import io.netty.handler.codec.stomp.StompCommand;
+import io.netty.handler.codec.stomp.StompFrame;
+
+import java.nio.CharBuffer;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static io.netty.buffer.ByteBufUtil.*;
+import static io.netty.handler.codec.stomp.StompHeaders.*;
+import static io.netty.util.CharsetUtil.*;
+
+@Sharable
+public class StompChatHandler extends SimpleChannelInboundHandler<StompFrame> {
+
+    static final StompChatHandler INSTANCE = new StompChatHandler();
+
+    private final ConcurrentHashMap<String, Set<StompSubscription>> CHAT_DESTINATIONS =
```

Comment:
wouldn't it be better to make this non static and create one instance of the handler that is shared at the end ? This way we ensure at least stuff can be GC'ed.

Reply:
Do you mean something like this:
```
@Sharable
public class StompWebSocketProtocolCodec extends MessageToMessageCodec<WebSocketFrame, ByteBuf> {

    private final StompChatHandler stompChatHandler = new StompChatHandler();

    @Override
    public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
        if (evt instanceof WebSocketServerProtocolHandler.HandshakeComplete) {
            StompVersion stompVersion = StompVersion.findBySubProtocol(((HandshakeComplete) evt).selectedSubprotocol());
            ctx.channel().attr(StompVersion.CHANNEL_ATTRIBUTE_KEY).set(stompVersion);
            String ctxName = ctx.name();
            ctx.pipeline()
               .addAfter(ctxName, "stompChatHandler", stompChatHandler)
               .addAfter(ctxName, "stompAggregator", new StompSubframeAggregator(65536))
               .addAfter(ctxName, "stompEncoder", new StompSubframeEncoder())
               .addAfter(ctxName, "stompDecoder", new StompSubframeDecoder())
               .remove(StompWebSocketClientPageHandler.INSTANCE);
        } else {
            super.userEventTriggered(ctx, evt);
        }
        super.userEventTriggered(ctx, evt);
    }
```

Reply:
yep... that said in your example you call `userEventTriggered` two times ;)

Reply:
Good caught 👍